### PR TITLE
[boost-container] Fix deps and emscripten

### LINKED ISSUE #38469: [boost-container] Build error on wasm32-emscripten
Package: boost-container:wasm32-emscripten@1.84.0#1

**Host Environment**

- Host: x64-windows
- Compiler: Clang 19.0.0
-    vcpkg-tool version: 2024-04-23-d6945642ee5c3076addd1a42c331bbf4cfc97457
    vcpkg-scripts version: 31de9445b 2024-04-30 (21 minutes ago)

**To Reproduce**

`vcpkg install `

**Failure logs**

```
-- Using cached boostorg-container-boost-1.84.0.tar.gz.
-- Cleaning sources at E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source E:/sandbox/toolchain/vcpkg/downloads/boostorg-container-boost-1.84.0.tar.gz
-- Using source at E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean
-- Configuring wasm32-emscripten
-- Building wasm32-emscripten-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j33
    Working Directory: E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/wasm32-emscripten-dbg
    See logs for more information:
      E:\sandbox\toolchain\vcpkg\buildtrees\boost-container\install-wasm32-emscripten-dbg-out.log

Call Stack (most recent call first):
  E:/sandbox/_obj-lib-etc-em/vcpkg_installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  E:/sandbox/_obj-lib-etc-em/vcpkg_installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  E:/sandbox/_obj-lib-etc-em/vcpkg_installed/x64-windows/share/vcpkg-boost/boost-install.cmake:65 (vcpkg_cmake_install)
  ports/boost-container/portfile.cmake:12 (boost_configure_and_install)
  scripts/ports.cmake:175 (include)



```

<details><summary>E:\sandbox\toolchain\vcpkg\buildtrees\boost-container\install-wasm32-emscripten-dbg-out.log</summary>

```
Change Dir: 'E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/wasm32-emscripten-dbg'

Run Build Command(s): E:/sandbox/toolchain/vcpkg/downloads/tools/ninja/1.10.2-windows/ninja.exe -v -v -j33 install
[1/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/dlmalloc.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\dlmalloc.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/dlmalloc.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/dlmalloc.cpp
[2/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/synchronized_pool_resource.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\synchronized_pool_resource.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/synchronized_pool_resource.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/synchronized_pool_resource.cpp
FAILED: libs/container/CMakeFiles/boost_container.dir/src/synchronized_pool_resource.cpp.o 
E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/synchronized_pool_resource.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\synchronized_pool_resource.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/synchronized_pool_resource.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/synchronized_pool_resource.cpp
In file included from E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/synchronized_pool_resource.cpp:14:
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:109:12: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes
  109 | extern "C" __declspec(dllimport) void __stdcall InitializeCriticalSection(::_RTL_CRITICAL_SECTION *);
      |            ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:109:39: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes]
  109 | extern "C" __declspec(dllimport) void __stdcall InitializeCriticalSection(::_RTL_CRITICAL_SECTION *);
      |                                       ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:111:12: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes
  111 | extern "C" __declspec(dllimport) void __stdcall EnterCriticalSection(::_RTL_CRITICAL_SECTION *);
      |            ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:111:39: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes]
  111 | extern "C" __declspec(dllimport) void __stdcall EnterCriticalSection(::_RTL_CRITICAL_SECTION *);
      |                                       ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:112:12: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes
  112 | extern "C" __declspec(dllimport) void __stdcall LeaveCriticalSection(::_RTL_CRITICAL_SECTION *);
      |            ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:112:39: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes]
  112 | extern "C" __declspec(dllimport) void __stdcall LeaveCriticalSection(::_RTL_CRITICAL_SECTION *);
      |                                       ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:113:12: error: '__declspec' attributes are not enabled; use '-fdeclspec' or '-fms-extensions' to enable support for __declspec attributes
  113 | extern "C" __declspec(dllimport) void __stdcall DeleteCriticalSection(::_RTL_CRITICAL_SECTION *);
      |            ^
E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include\boost/container/detail/thread_mutex.hpp:113:39: warning: '__stdcall' calling convention is not supported for this target [-Wignored-attributes]
  113 | extern "C" __declspec(dllimport) void __stdcall DeleteCriticalSection(::_RTL_CRITICAL_SECTION *);
      |                                       ^
4 warnings and 4 errors generated.
[3/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/unsynchronized_pool_resource.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\unsynchronized_pool_resource.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/unsynchronized_pool_resource.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/unsynchronized_pool_resource.cpp
[4/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/monotonic_buffer_resource.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\monotonic_buffer_resource.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/monotonic_buffer_resource.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/monotonic_buffer_resource.cpp
[5/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/pool_resource.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\pool_resource.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/pool_resource.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/pool_resource.cpp
[6/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\em++.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -std=gnu++11 -fvisibility=hidden -fvisibility-inlines-hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/global_resource.cpp.o -MF libs\container\CMakeFiles\boost_container.dir\src\global_resource.cpp.o.d -o libs/container/CMakeFiles/boost_container.dir/src/global_resource.cpp.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/global_resource.cpp
[7/9] E:\sandbox\toolchain\emsdk\upstream\emscripten\emcc.bat -DBOOST_CONTAINER_NO_LIB -DBOOST_CONTAINER_STATIC_LINK -IE:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/include -isystem E:/sandbox/_obj-lib-etc-em/vcpkg_installed/wasm32-emscripten/include -g -fvisibility=hidden -MD -MT libs/container/CMakeFiles/boost_container.dir/src/alloc_lib.c.o -MF libs\container\CMakeFiles\boost_container.dir\src\alloc_lib.c.o.d -o libs/container/CMakeFiles/boost_container.dir/src/alloc_lib.c.o -c E:/sandbox/toolchain/vcpkg/buildtrees/boost-container/src/ost-1.84.0-f9f7c1b921.clean/libs/container/src/alloc_lib.c
ninja: build stopped: subcommand failed.
```
</details>

**Additional context**

<details><summary>vcpkg.json</summary>

```
{
  "name": "engine",
  "version": "1.0.0",
  "dependencies": [
    "boost-container"
  ]
}

```
</details>

### LINKED ISSUE #38679: [boost-container] build failure on Windows (missing boost-static-assert dependency)
### Operating system

Windows

### Compiler

MSVC

### Steps to reproduce the behavior

```Shell
`vcpkg install boost-container`
```

### Failure logs

```
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: "C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe" -v
    Working Directory: C:/vcpkg/buildtrees/boost-container/x64-windows-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      C:\vcpkg\buildtrees\boost-container\config-x64-windows-dbg-CMakeCache.txt.log
      C:\vcpkg\buildtrees\boost-container\config-x64-windows-rel-CMakeCache.txt.log
      C:\vcpkg\buildtrees\boost-container\config-x64-windows-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:261 (vcpkg_execute_required_process)
  installed/x64-windows/share/vcpkg-boost/boost-install.cmake:55 (vcpkg_cmake_configure)
  ports/boost-container/portfile.cmake:12 (boost_configure_and_install)
  scripts/ports.cmake:175 (include)
```

C:\vcpkg\buildtrees\boost-container\config-x64-windows-out.log:

```
  Could not find a package configuration file provided by
  "boost_static_assert" (requested version 1.85.0) with any of the following
  names:

    boost_static_assertConfig.cmake
    boost_static_assert-config.cmake

  Add the installation prefix of "boost_static_assert" to CMAKE_PREFIX_PATH
  or set "boost_static_assert_DIR" to a directory containing one of the above
  files.  If "boost_static_assert" provides a separate development package or
  SDK, be sure it has been installed.
Call Stack (most recent call first):
  C:/vcpkg/installed/x64-windows/share/boost/BoostConfig.cmake:67 (find_package)
  C:/vcpkg/installed/x64-windows/share/boost/BoostConfig.cmake:128 (boostcfg_find_component)
  C:/vcpkg/installed/x64-windows/share/boost/vcpkg-cmake-wrapper.cmake:3 (_find_package)
  C:/vcpkg/scripts/buildsystems/vcpkg.cmake:813 (include)
  C:/vcpkg/installed/x64-windows/share/boost/cmake-build/BoostRoot.cmake:234 (find_package)
  C:/vcpkg/installed/x64-windows/share/boost/cmake-build/BoostRoot.cmake:274 (__boost_scan_dependencies)
  CMakeLists.txt:10 (include)
```

### Additional context

Appears to be a missing dependency on the boost-static-assert package.

Running `vcpkg install boost-static-assert` before running `vcpkg install boost-container` fixes the issue.

### PR DESCRIPTION
**Pull request by dg0yt**
Fixes #38679. ([No longer uses Boost::static_assert.](https://www.boost.org/doc/libs/1_85_0/doc/html/container/release_notes.html#container.release_notes.release_notes_boost_1_85_00)) 
Fixes #38469. (Needs threads, so [emscripten needs to use `-pthread`](https://emscripten.org/docs/porting/pthreads.html#compiling-with-pthreads-enabled).)
Change homepage link to something more useful. The update to the generator script and the other ports will be in a separate PR.

### CODE REVIEW

**Comment by Neumann-A** [comment]
I think the correct approach is to patch the CMakeLists.txt and remove the dep. https://github.com/boostorg/container/pull/277

**Comment by dg0yt** [comment]
Sure, I had this already, now that I found it out. Do you also want to fix emscripten?

**Comment by Neumann-A** [comment]
What is wrong with emscripten?

**Comment by dg0yt** [comment]
What is wrong with `boost-container`?!

If it doesn't find POSIX threads (`BOOST_HAS_PTHREADS`), it blindly switches to Windows threads.
https://github.com/boostorg/container/blob/develop/include/boost/container/detail/thread_mutex.hpp#L38-L81

But `BOOST_HAS_PTHREADS` isn't getting true unless `_POSIX_THREADS` is true enough.
https://github.com/boostorg/config/blob/5734e160e08b8df898c7f747000f27a3aafb7b2b/include/boost/config/detail/posix_features.hpp#L45-L47

emscripten sets `_POSIX_THREADS` to `-1` if `-ptreads` isn't used.

File: ports/boost-container/emscripten-pthread.diff [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+   src/synchronized_pool_resource.cpp
+   src/unsynchronized_pool_resource.cpp
+ )
++if(EMSCRIPTEN)
++  # BOOST_HAS_PTHREADS is required to avoid windows declarations in
++  # libs/container/include/boost/container/detail/thread_mutex.hpp
++  # This should come from include/boost/config/detail/posix_features.hpp,
++  # but ... https://github.com/microsoft/vcpkg/issues/38469
++  target_compile_options(boost_container PUBLIC -pthread)
++endif()
```
Comment (Neumann-A) [inline ports/boost-container/emscripten-pthread.diff:15]: Why not:

```
find_package(Threads)
target_link_libraries(boost_container
  PUBLIC
  Threads::Threads
```

?

Comment (dg0yt) [reply ports/boost-container/emscripten-pthread.diff:15]: Just a minimal fix without adding `find_dependency`.

Comment (Neumann-A) [reply ports/boost-container/emscripten-pthread.diff:15]: You don't need to add it. Just add it to the `target_link_libraries` list, `boost-cmake` should take care of the rest.

Comment (Neumann-A) [reply ports/boost-container/emscripten-pthread.diff:15]: https://github.com/boostorg/cmake/blob/6dcde32c82f7c1fef67b6e61db862f73e4809210/include/BoostInstall.cmake#L343-L347

Comment (dg0yt) [reply ports/boost-container/emscripten-pthread.diff:15]: Unfortunately, `FindThreads.cmake` doesn't enable the required `-pthread` for emscripten. `THREADS_PREFER_PTHREAD_FLAG` has no effect once CMake found `CMAKE_HAVE_LIBC_PTHREAD`, based on a pure compile check.

**Review by MonicaLiu0311 — APPROVED** [review]

**Comment by rjindael** [comment]
I would like to add that issue https://github.com/microsoft/vcpkg/issues/38469 (which this PR solves) is currently breaking installing boost altogether (if you include only "boost" in your vcpkg) on the wasm32-emscripten configuration so ideally this PR should be merged soon.

**Comment by dg0yt** [comment]
> installing boost altogether 

... isn't what I would recommend.

It is no surprise that the meta port stumbles over each failing dependency. After `boost-container` (this PR) comes `boost-interprocess` (#38815).

**Pull request merged**